File: mobile-sdk/flutter-sdk/lib/utils/retry.dart
import 'dart:async';
import 'dart:math';
import '../types/types.dart';

typedef RetryCallback<T> = Future<T> Function();
typedef ShouldRetryCallback = bool Function(PyAirtableException error);
typedef OnRetryCallback = void Function(PyAirtableException error, int attempt);

class RetryOptions {
  final int attempts;
  final int delay;
  final ShouldRetryCallback? shouldRetry;
  final OnRetryCallback? onRetry;

  const RetryOptions({
    required this.attempts,
    required this.delay,
    this.shouldRetry,
    this.onRetry,
  });
}

/// Retry a function with exponential backoff
Future<T> withRetry<T>(
  RetryCallback<T> callback,
  RetryOptions options,
) async {
  PyAirtableException? lastError;
  
  for (int attempt = 1; attempt <= options.attempts; attempt++) {
    try {
      return await callback();
    } catch (error) {
      lastError = error is PyAirtableException
          ? error
          : PyAirtableException(error.toString());
      
      // Don't retry on last attempt or if error is not retryable
      if (attempt == options.attempts || 
          (options.shouldRetry != null && !options.shouldRetry!(lastError))) {
        throw lastError;
      }
      
      options.onRetry?.call(lastError, attempt);
      
      // Wait before retrying
      final retryDelay = getRetryDelay(attempt, options.delay);
      await Future.delayed(Duration(milliseconds: retryDelay));
    }
  }
  
  throw lastError!;
}

/// Check if error is retryable
bool isRetryableError(PyAirtableException error) {
  if (error is NetworkException) return true;
  if (error is ServerException) return true;
  if (error is RateLimitException) return true;
  
  // Retry on specific status codes
  if (error.statusCode != null) {
    return [
      HttpStatus.badGateway,
      HttpStatus.serviceUnavailable,
      HttpStatus.tooManyRequests,
    ].contains(error.statusCode);
  }
  
  return false;
}

/// Get retry delay with exponential backoff and jitter
int getRetryDelay(int attempt, int baseDelay) {
  // Exponential backoff with jitter
  final exponentialDelay = baseDelay * pow(2, attempt - 1).toInt();
  final jitter = (Random().nextDouble() * 0.1 * exponentialDelay).toInt();
  return min(exponentialDelay + jitter, 30000); // Max 30 seconds
}

/// Create a timeout future that completes with an error after a given time
Future<T> timeout<T>(Future<T> future, int milliseconds) {
  return future.timeout(
    Duration(milliseconds: milliseconds),
    onTimeout: () => throw const PyAirtableException('Request timeout'),
  );
}

/// Debounce function calls
class Debouncer {
  final int milliseconds;
  Timer? _timer;

  Debouncer({required this.milliseconds});

  void run(VoidCallback action) {
    _timer?.cancel();
    _timer = Timer(Duration(milliseconds: milliseconds), action);
  }

  void dispose() {
    _timer?.cancel();
  }
}

/// Throttle function calls
class Throttler {
  final int milliseconds;
  bool _isThrottled = false;

  Throttler({required this.milliseconds});

  void run(VoidCallback action) {
    if (!_isThrottled) {
      action();
      _isThrottled = true;
      Timer(Duration(milliseconds: milliseconds), () {
        _isThrottled = false;
      });
    }
  }
}

typedef VoidCallback = void Function();

File: mobile-sdk/react-native-sdk/src/utils/retry.ts
import { PyAirtableSDKError, isRetryableError, getRetryDelay } from './errors';

export interface RetryOptions {
  attempts: number;
  delay: number;
  shouldRetry?: (error: PyAirtableSDKError) => boolean;
  onRetry?: (error: PyAirtableSDKError, attempt: number) => void;
}

/**
 * Retry a function with exponential backoff
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  options: RetryOptions
): Promise<T> {
  const { attempts, delay, shouldRetry = isRetryableError, onRetry } = options;
  
  let lastError: PyAirtableSDKError;
  
  for (let attempt = 1; attempt <= attempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error instanceof PyAirtableSDKError ? error : 
        new PyAirtableSDKError(error.message || 'Unknown error');
      
      // Don't retry on last attempt or if error is not retryable
      if (attempt === attempts || !shouldRetry(lastError)) {
        throw lastError;
      }
      
      onRetry?.(lastError, attempt);
      
      // Wait before retrying
      const retryDelay = getRetryDelay(attempt, delay);
      await sleep(retryDelay);
    }
  }
  
  throw lastError!;
}

/**
 * Sleep for a given number of milliseconds
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Create a timeout promise that rejects after a given time
 */
export function timeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return Promise.race([
    promise,
    new Promise<never>((_, reject) =>
      setTimeout(() => reject(new PyAirtableSDKError('Request timeout')), ms)
    ),
  ]);
}

/**
 * Debounce a function to prevent multiple rapid calls
 */
export function debounce<T extends (...args: any[]) => any>(
  func: T,
  wait: number
): (...args: Parameters<T>) => void {
  let timeoutId: NodeJS.Timeout | null = null;
  
  return (...args: Parameters<T>) => {
    if (timeoutId) {
      clearTimeout(timeoutId);
    }
    
    timeoutId = setTimeout(() => {
      func(...args);
    }, wait);
  };
}

/**
 * Throttle a function to limit execution frequency
 */
export function throttle<T extends (...args: any[]) => any>(
  func: T,
  limit: number
): (...args: Parameters<T>) => void {
  let inThrottle = false;
  
  return (...args: Parameters<T>) => {
    if (!inThrottle) {
      func(...args);
      inThrottle = true;
      setTimeout(() => inThrottle = false, limit);
    }
  };
}